### ISSUE TITLE
Parallelize prune pack rewrites

### ISSUE THREAD

**Issue by pmkane** [restic/restic#1470]
Currently, when restic rewrites packs as part of a purge, the process is single threaded.

In an Amazon S3 environment, this results in long prune times.  For example, in our ~850k pack repository where we keep 7 daily backups from several machines, about 35k packs need to be rewritten each day.  This takes approximately 3.5 hours with a single thread.

Parallelizing this would significantly decrease the amount of time required to complete a forget/purge cycle.

**Comment by fd0** [comment]
Agreed, good suggestion, thanks!

**Comment by pmkane** [comment]
Just an additional data point -- a repository with 399,318 packs (2,945,147 blobs) with 719,021 duplicate blobs (403.384 GiB duplicate) -- has been in the "rewriting packs" phase of prune for 39 hours and is only 54.22% done.

Host running restic is beefy (i3.xlarge) and is in the same AZ as the S3 bucket it is backing up to, so latency/throughput isn't the issue.

**Comment by fd0** [comment]
I'll see what I can do short-term.

**Comment by gransear** [comment]
This looks like the same problem as I had in #1348. I have a large latency so it will take a week to prune.

```
will delete 6668 packs and rewrite 29873 packs, this frees 103.002 GiB
[15:40:42] 7.85%  2344 / 29873 packs rewritten
```

**Comment by fd0** [comment]
@mjdsys @ifedorenko Since you've done some great PRs in the last couple of weeks, and you seem interested in improving restic's performance, is one of you maybe interested in giving this a try (with my help)?

**Comment by MJDSys** [comment]
@fd0 I'd be interested in helping out some, though it may take a bit for me to polish it all up.  @ifedorenko if you get some time ahead of me, feel free to pick this up instead.

In terms of what to do, taking a quick look at the code I see:

1. In repository/repacker_manager.go: This needs to be adjusted to ensure it can be called from multiple threads simultaneously without problems.  I assume since the backup does do work in parallel, this might already be prepared.  But I've verified this so I put this here just to make sure.
1. In repository/repack.go, when handling each pack instead of saving each blob as we come to them, instead these should be handed off to some worker pool of go routines.  Initially, each blob can just be handed off randomly through a channel.
1.  Finally, benchmarking this entire system, ensuring that communication overhead is kept to a minimum.  Depending upon blob size, this might require handling off a set of blobs to each worker thread.  Potentially also processing the list of packs in parallel might help as well?

I think with some contexts this can be easily cancelled on errors (or even allowed to proceed as far as possible before returning all the errors, whichever is preferred).  @fd0, if you could confirm is packer_manager is already threadsafe, I can start working on a WIP PR for 2) while we decide on 3).  How does that plan sound?

**Comment by ifedorenko** [comment]
I have not looked at purge implementation yet, so I am unlikely to have anything meaningful to say or contribute for couple of weeks at least (and I still need to finish `List()` related changes and `Repository.ListPack()` optimization we discussed in #1567). Purge is the next on my TODO list, but I was planning to think about "lock-free purge" approach before doing any code changes (relying on lock files is kinda scary for eventually-consistent backends like S3 or OneDrive, if you think about it).

**Comment by stooone** [comment]
Maybe it's not for this issue, but it would be nice if prune could continue if interrupted. My home PC isn't up 7x24 and I backup around 300G and 250.000 files every day (to B2). Last time I tried to purge I interrupted it after two days... (using cache) Then I redid my backup from scratch instead, since it continues after an interrupt and was faster.

**Comment by MJDSys** [comment]
@ifedorenko Ok, then I'll leave this alone for now.  Or think some on the lock free approach too.

**Comment by ifedorenko** [comment]
So I am looking at prune performance now. Couple of question to @fd0 or anyone else who knows the background

* If I read the code correctly, prune rewrites packs with unused blobs regardless how much space this actually saves. Have you considered less aggressive repack strategy, which only rewrites packs that are >10% unused (or some other, possibly configurable, threshold)?
* Is it necessary to decrypt and re-encrypt blobs during pack rewrite? I thought pack format allowed moving encrypted blobs around.

**Comment by MJDSys** [comment]
> If I read the code correctly, prune rewrites packs with unused blobs regardless how much space this actually saves.
From my brief look through the code, this does seem to be true.  One idea I had was to make the packs (or individual blobs in a pack) have a generation number, and try to pack files that have lasted in the repository for longer periods of time together, instead of randomly packing together files as they are found.  This would be similar in idea to a generational garbage collector.

**Comment by ifedorenko** [comment]
Yes, agree, "generations garbage collection" is another possible way to optimize pack rewrite, but I think it is orthogonal to how aggressively rewrite compacts unused space. In both cases we'll likely want to instrument the code to collect stats first so we can assess how much benefit (if any) such optimizations will bring.

**Comment by fd0** [comment]
> If I read the code correctly, prune rewrites packs with unused blobs regardless how much space this actually saves. Have you considered less aggressive repack strategy, which only rewrites packs that are >10% unused (or some other, possibly configurable, threshold)?

Yes, I like that strategy and I think making the amount of space that is "wasted" configurable for the end user will cover a whole range of use cases, e.g. between 0% (current strategy, repack all files which have at least one unused blob in it) and 100% (only remove pack files which are completely unused). Can be implemented without much code I think.

> Is it necessary to decrypt and re-encrypt blobs during pack rewrite? I thought pack format allowed moving encrypted blobs around.

It isn't strictly necessary, we could just re-upload the same blob. Each blob should be decrypted during `prune` though, to check if it contains the right data (and the MAC matches, so that we can be sure it's unmodified).

> One idea I had was to make the packs (or individual blobs in a pack) have a generation number, and try to pack files that have lasted in the repository for longer periods of time together, instead of randomly packing together files as they are found. This would be similar in idea to a generational garbage collector.

That's a nice idea which we should save for later, as it requires modifying the repository format.

So, the way forward is twofold:
 * Make `prune` configurable which pack files are rewritten (e.g. as a percentage)
 * Make `prune` rewrite packs in parallel

**Comment by ifedorenko** [comment]
> It isn't strictly necessary, we could just re-upload the same blob. Each blob should be decrypted during prune though, to check if it contains the right data (and the MAC matches, so that we can be sure it's unmodified).

I _think_ this conflates `prune` and `check` and may be confusing to the users (e.g. #1596). But, regardless, I just wanted to make sure I understood repository format, I do not propose we change current behaviour. 

Back to "parallel prune". Unless somebody else is already working on this, I plan to implement configurably less aggressive repack and parallel repack in the next few days, time permitting. 

Couple of other ideas for discussion:

* Provide an option to reusing existing index during repack. Currently repack rebuilds repository index twice, before and after pack rewrite. For "~850k pack S3 repository" this likely takes 30+ minutes even with recent `index.New` optimizations. I think time saving justifies additional complexity of incremental index maintenance.
* Instrument the code to collect stats and upload them to the repository. Currently we have to guess where the time is spent and more importantly why. I think knowing exactly how many packs gets rewritten, how much space is reclaimed with each pack, types of S3 requests and their respective timings, etc, will help us make more informed decisions how to optimize prune.

What do you think?

**Comment by fd0** [comment]
> Back to "parallel prune". Unless somebody else is already working on this, I plan to implement configurably less aggressive repack and parallel repack in the next few days, time permitting.

Awesome, I'm looking forward to it!

About your ideas:

> Provide an option to reusing existing index during repack. Currently repack rebuilds repository index twice, before and after pack rewrite. For "~850k pack S3 repository" this likely takes 30+ minutes even with recent index.New optimizations. I think time saving justifies additional complexity of incremental index maintenance.

Yes, that's possible and a good thing to do. We can use the old index to get information about existing pack files, we can even estimate that the index makes sense (sum of blob bytes equals pack file size etc).

I don't know if you've seen it, but at the moment we've got two different "index" implementations, in `internal/restic/index` and `internal/restic/repository`... I started reworking index handling and got sidetracked. The former is the newer implementation, but the repository still uses the older (which is loaded from the repo and not rebuilt), which leads to confusing behavior such as the panic described in https://forum.restic.net/t/error-loading-tree-check-prune-and-forget-gives-error-b2-backend/406?u=fd0

> Instrument the code to collect stats and upload them to the repository. Currently we have to guess where the time is spent and more importantly why. I think knowing exactly how many packs gets rewritten, how much space is reclaimed with each pack, types of S3 requests and their respective timings, etc, will help us make more informed decisions how to optimize prune.

I'm not so sure about that, I don't like storing this in the repo. Where would you store it exactly? I don't like modifying the repo format... What about saving it to a JSON file on request and nudge people to submit these?

**Comment by ifedorenko** [comment]
I was thinking saving the stats to `stats/` directory in the repository (sibling to `data/`, `index/`, etc). This way there is single place to look for the stats, even when multiple clients use the same repository, or when tasks like `check` or `prune` run on ephemeral VMs. Don't have a strong opinion, however, local filesystem is likely a good start too (still need to decide local filesystem location).

**Comment by teknynja** [comment]
I'm looking over the code to get an idea of the complexity/feasibility of implementing "adjustable aggressiveness" of repacking blobs, and I think I have a fairly good idea of what's happening in the code, but I'm confused about the purpose of adding `mixedBlobs` (packs with both tree and data blobs) to `rewritePacks` collection [see `cmd/restic/cmd_prune.go`:217]. If a pack has mixed blobs but all the data blobs are used/active, why do we need to re-write that pack? Conversely, if a pack has all tree blobs, why don't we need to re-write the pack?

It seems like a lot of packs would meet the `mixedBlobs` critiera (especially when backing up a lot of smaller files) which means a lot of packs may get re-written. I'm still getting my head around how restic works under the hood (and learning go along the way) and I'm probably missing something basic, so if somebody could point out the reasoning behind the mixed blobs handling it would help a lot in my understanding of what's going on.

For some background, I'm using restic with the b2 backend, and I'm finding pruning costs a lot in terms of transactions (and therefore money) and I'm looking for a way to balance the cost of removing data with the cost of storing it. Because many of my files are large (virtual machine disk images) I'm thinking that my changes are over large areas of those files and thus forgetting snapshots will result in a large number of fully unused packs that can simply be deleted to save data, without needing to go through the processes of repacking every pack with even a single unused blob.

@ifedorenko - You mentioned you were looking into dealing with the adjustable repacking aggressiveness as well, have you had a chance to dig into this as well?

**Comment by ifedorenko** [comment]
No, I am not looking at prune/repack at the moment. Already have too many local changes, want to see the new restorer merged before I consider any other non-trivial restic work.

**Comment by fd0** [comment]
I can clear up what's going on :)

Originally, the restic repository specification allows pack files to contain both types of blobs, `data` and `tree`. We did not change that for the sake of backwards-compatibility, so older versions of restic will work just fine.

In 0.8.0 we added the metadata cache, which automatically downloads and caches pack files containing only tree blobs. At the same time we also added code so that pack files only contain `data` or `tree` blobs, and the code you discovered in the `prune` command which re-packs mixed blobs.

So, from the next run of `prune` all pack files will contain either `tree` blobs (and can be cached) or `data` blobs (so they won't be cached).

That's the reason for the code :)

**Comment by teknynja** [comment]
Oh, that makes sense now! Thank you. It seems that quite a few things are out-of-date in the [restic Design Document](https://restic.readthedocs.io/en/v0.2.0/Design/) but that was something I couldn't figure out.

With that "mystery" solved I think I'm ready to begin playing with some ideas for adjusting the aggressiveness of repacking.

Thanks again for the input.

**Comment by fd0** [comment]
Hm, besides mentioning that by default we now have data/tree-only pack files, is there anything else we can improve in the design document?

**Comment by teknynja** [comment]
There were mostly minor issues with the example data structures not matching what I'm actually seeing in my repository, for example the snapshot example has `dir`, `uid`, & `gid` fields where I'm finding `paths`, `excludes`, `parent` fields in the actual dump - it's easy enough to figure out what has changed there.

What proved to be a bit confusing was was the example command `restic cat tree`. Apparently the `tree` type had been removed, and it wasn't clear at first because when I tried `restic cat tree ID` instead of telling me that `tree` is an invalid type, it gives the message `Fatal: invalid type` which I was thinking meant the repository data was invalid, not my command. (Arrogance on my part? :P)

Anyway, I finally figured out the command was supposed to be `restic cat blob` and was able to view the tree record. Unfortunately, it is hard to view because the JSON is condensed. There is an example in the "Keys, Encryption and MAC" section mentioning how to view pretty-printed JSON. I worked out the command to view the tree data in human form using `restic -q cat blob ID | python -m json.tool`. It looks like there the only field not depicted in the design document (for both tree and data nodes) are the `device_id` fields, which wasn't wasn't important to me trying to decipher the structure.

I didn't dig into the key and index structures, so I'm not sure how closely the document matches the current structures now.

Overall the design document is incredibly helpful, as I don't think I would have been able to figure out what the code was doing without it - more projects should consider having this level of detail outlining the major data structures used in their programs.

**Comment by teknynja** [comment]
@fd0 I'd like to discuss some implementation options - the code I have now was not intended to be committed, just something to play around with ideas and learn about the code (and go) a bit.

I primarily wanted to get some input on how the command line flag should be implemented, but there are a few other questions as well.

**Comment by aawsome** [comment]
I think this issue is closed by #2941

**Comment by fd0** [comment]
It is, thanks for the reminder!
